rcl/src/rcl/node_type_cache.c

```diff
@@ -0,0 +1,254 @@
+}
+
+static rcl_ret_t rcl_node_type_cache_register_type_info(
+    const rcl_node_t *node, const rosidl_type_hash_t *type_hash,
+    const rosidl_runtime_c__type_description__TypeDescription
+        *type_description) {
+  char *type_hash_str = NULL;
+  rcl_type_info_t type_info;
+
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_hash, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_description, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(node->impl->registered_types_by_type_hash,
+                              "type cache not initialized",
+                              return RCL_RET_NOT_INIT);
+
+  // Convert hash to string
+  if (RCUTILS_RET_OK !=
+      rosidl_stringify_type_hash(type_hash, rcutils_get_default_allocator(),
+                                 &type_hash_str)) {
+    RCL_SET_ERROR_MSG("Failed to stringify type hash");
+    return RCL_RET_ERROR;
+  }
+
+  // TODO(achim-k): Remove log statement
+  const char *typeName = type_description->type_description.type_name.data;
+  RCUTILS_LOG_INFO_NAMED(ROS_PACKAGE_NAME, "R: %s (%s)", type_hash_str,
+                         typeName);
+
+  // If the type already exists, we only have to increment the registration
+  // count.
+  if (rcutils_hash_map_key_exists(node->impl->registered_types_by_type_hash,
+                                  &type_hash_str)) {
+    if (RCUTILS_RET_OK !=
+        rcutils_hash_map_get(node->impl->registered_types_by_type_hash,
+                             &type_hash_str, &type_info)) {
+      RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Failed to retrieve type info for type '%s'", type_hash_str);
+      return RCL_RET_ERROR;
+    }
+
+    type_info.numRegistrations++;
+  } else {
+    // First registration of this type
+    type_info.numRegistrations = 1;
+
+    // Convert type description struct to type description message struct.
+    type_info.type_description =
+        rcl_convert_type_description_runtime_to_msg(type_description);
+    RCL_CHECK_FOR_NULL_WITH_MSG(type_info.type_description,
+                                "converting type description struct failed",
+                                return RCL_RET_ERROR);
+  }
+
+  // Update the hash map entry.
+  if (RCUTILS_RET_OK !=
+      rcutils_hash_map_set(node->impl->registered_types_by_type_hash,
+                           &type_hash_str, &type_info)) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Failed to update type info for type '%s'", type_hash_str);
+    return RCL_RET_ERROR;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t rcl_node_type_cache_init(const rcl_node_t *node) {
+  RCL_CHECK_ARGUMENT_FOR_NULL(node, RCL_RET_INVALID_ARGUMENT);
+
+  if (NULL != node->impl->registered_types_by_type_hash) {
+    RCL_SET_ERROR_MSG("Type cache already initialized");
+    return RCL_RET_ALREADY_INIT;
+  }
+
+  node->impl->registered_types_by_type_hash =
+      node->context->impl->allocator.allocate(
```

Comment:
Note: when using direct struct on the impl instead of pointer, no `allocate` needed here